tests/protocol/mocks.py

```diff
@@ -0,0 +1,142 @@
+          methods stop_mock_http_get() and stop_mock_crypt_util().
+
+    The return value is an instance of WireProtocol augmented with these properties/methods:
+
+        * mock_wire_data - the WireProtocolData constructed from the mock_wire_data_file parameter.
+        * stop_mock_http_request() - stops the mock for restutil.http_request
+        * stop_mock_crypt_util() - stops the mock for CrypUtil
+        * stop() - stops both mocks
+    """
+    def stop_mock_http_request():
+        if stop_mock_http_request.mock is not None:
+            stop_mock_http_request.mock.stop()
+            stop_mock_http_request.mock = None
+    stop_mock_http_request.mock = None
+
+    def stop_mock_crypt_util():
+        if stop_mock_crypt_util.mock is not None:
+            stop_mock_crypt_util.mock.stop()
+            stop_mock_crypt_util.mock = None
+    stop_mock_crypt_util.mock = None
+
+    def stop():
+        stop_mock_crypt_util()
+        stop_mock_http_request()
+
+    protocol = WireProtocol(restutil.KNOWN_WIRESERVER_IP)
+    protocol.mock_wire_data = WireProtocolData(mock_wire_data_file)
+    protocol.stop_mock_http_request = stop_mock_http_request
+    protocol.stop_mock_crypt_util = stop_mock_crypt_util
+    protocol.stop = stop
+
+    try:
+        # To minimize the impact of mocking restutil.http_request we only use the mock data for requests
+        # to the wireserver or requests starting with "mock-goal-state"
+        mock_data_re = re.compile(r'https?://(mock-goal-state|{0}).*'.format(restutil.KNOWN_WIRESERVER_IP.replace(r'.', r'\.')), re.IGNORECASE)
+
+        original_http_request = restutil.http_request
+
+        def http_request(method, url, data, headers=None, use_proxy=False, max_retry=restutil.DEFAULT_RETRIES, retry_codes=restutil.RETRY_CODES, retry_delay=restutil.DELAY_IN_SECONDS):
+            if method == 'GET' and mock_data_re.match(url) is not None:
+                return protocol.mock_wire_data.mock_http_get(url, headers, use_proxy, max_retry, retry_codes, retry_delay)
+            elif method == 'POST':
+                return protocol.mock_wire_data.mock_http_post(url, data, headers, use_proxy, max_retry, retry_codes, retry_delay)
+            return original_http_request(method, url, data, headers, use_proxy, max_retry, retry_codes, retry_delay)
+
+        patched = patch("azurelinuxagent.common.utils.restutil.http_request", side_effect=http_request)
+        patched.start()
+        stop_mock_http_request.mock = patched
+
+        patched = patch("azurelinuxagent.common.protocol.wire.CryptUtil", side_effect=protocol.mock_wire_data.mock_crypt_util)
+        patched.start()
+        stop_mock_crypt_util.mock = patched
+
+        protocol.detect()
+
+        yield protocol
+
+    finally:
+        protocol.stop()
+
+
+@contextlib.contextmanager
+def mock_http_request(http_get_handler=None, http_post_handler=None, http_put_handler=None):
```

Comment:
New mock useful for tests that need to interact with the host/storage at the HTTP request level